ekm/remote_key_manager.go

```diff
@@ -0,0 +1,417 @@
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	ssz "github.com/ferranbt/fastssz"
+	slashingprotection "github.com/ssvlabs/eth2-key-manager/slashing_protection"
+	spectypes "github.com/ssvlabs/ssv-spec/types"
+	"go.uber.org/zap"
+
+	ssvsignerclient "github.com/ssvlabs/ssv/ssvsigner/client"
+	"github.com/ssvlabs/ssv/ssvsigner/web3signer"
+
+	"github.com/ssvlabs/ssv/beacon/goclient"
+	"github.com/ssvlabs/ssv/networkconfig"
+	"github.com/ssvlabs/ssv/operator/keys"
+	ssvtypes "github.com/ssvlabs/ssv/protocol/v2/types"
+	"github.com/ssvlabs/ssv/storage/basedb"
+)
+
+type RemoteKeyManager struct {
+	logger          *zap.Logger
+	client          *ssvsignerclient.SSVSignerClient
+	consensusClient *goclient.GoClient
+	getOperatorId   func() spectypes.OperatorID
+	retryCount      int
+	operatorPubKey  keys.OperatorPublicKey
+	SlashingProtector
+}
+
+func NewRemoteKeyManager(
+	logger *zap.Logger,
+	client *ssvsignerclient.SSVSignerClient,
+	consensusClient *goclient.GoClient,
+	db basedb.Database,
+	networkConfig networkconfig.NetworkConfig,
+	getOperatorId func() spectypes.OperatorID,
+	options ...Option,
+) (*RemoteKeyManager, error) {
+
+	signerStore := NewSignerStorage(db, networkConfig.Beacon, logger)
+	protection := slashingprotection.NewNormalProtection(signerStore)
+
+	sp, err := NewSlashingProtector(logger, signerStore, protection)
+	if err != nil {
+		logger.Fatal("could not create new slashing protector", zap.Error(err))
+	}
+
+	operatorPubKeyString, err := client.GetOperatorIdentity()
+	if err != nil {
+		return nil, fmt.Errorf("get operator identity: %w", err)
+	}
+
+	operatorPubKey, err := keys.PublicKeyFromString(operatorPubKeyString)
+	if err != nil {
+		return nil, fmt.Errorf("extract operator public key: %w", err)
+	}
+
+	s := &RemoteKeyManager{
+		logger:            logger,
+		client:            client,
+		consensusClient:   consensusClient,
+		SlashingProtector: sp,
+		getOperatorId:     getOperatorId,
+		operatorPubKey:    operatorPubKey,
+	}
+
+	for _, option := range options {
+		option(s)
+	}
+
+	return s, nil
+}
+
+type Option func(signer *RemoteKeyManager)
+
+func WithLogger(logger *zap.Logger) Option {
+	return func(s *RemoteKeyManager) {
+		s.logger = logger.Named("remote_key_manager")
+	}
+}
+
+func WithRetryCount(n int) Option {
+	return func(s *RemoteKeyManager) {
+		s.retryCount = n
+	}
+}
+
+func (km *RemoteKeyManager) AddShare(encryptedShare []byte) error {
+	var statuses []ssvsignerclient.Status
+	var publicKeys [][]byte
+	f := func() error {
+		var err error
+		statuses, publicKeys, err = km.client.AddValidators(encryptedShare)
+		return err
+	}
+	err := km.retryFunc(f, "AddValidators")
+	if err != nil {
+		return fmt.Errorf("add validator: %w", err)
+	}
+
+	if statuses[0] == ssvsignerclient.StatusImported || statuses[0] == ssvsignerclient.StatusDuplicated {
+		if err := km.BumpSlashingProtection(publicKeys[0]); err != nil {
+			return fmt.Errorf("could not bump slashing protection: %w", err)
+		}
```

Comment:
shouldn't you be doing this for all the publicKeys?

Reply:
ahh I see this is a single Share.

Reply:
IMO, no. Web3signer supports adding multiple keys at once, so I think the web3signer client should also add several keys. However, AddShare passes just one share, so it always receives one response. The web3signer checks that the response length is equal to the request length.